### DISCUSSION TITLE
Enable switching of password hash algorithm between bcrypt and Argon2

### DISCUSSION THREAD

**Discussion by x-EricH-x** [directus/directus#4718] [Feature Requests]
Since bcrypt is a commonly used hash algorithm, and especially, it is used in Directus v8, I suggest having:

1. an option to select the password hash algorithm
2. alternatively, an automatic re-hash of password using Argon2 upon login if the password is hashed using bcrypt

I think (2) is a better option because (1) allows user to stay using bcrypt, which defeat the intention of changing to a more secure hash algorithm, Argon2.

Use case are for people upgrading from Directus v8 or people trying to migrate from other CMS.

**Comment by joeinnes** [comment]
I went down a rabbit hole on this before, so here's what I found consolidated:

**Conclusion 1: Only Argon2 hashed passwords will work in Directus**
* https://github.com/directus/directus/blob/2d59412209f92efd00ea2f34a50dbf6d3660b39c/api/src/services/authentication.ts#L89, 

**Conclusion 2: The v8 migration docs are incomplete** 
They imply that you can just run the script and your Directus will be upgraded. It is true that the users are migrated, but they will not be able to log in without resetting their password)
* https://v8.docs.directus.io/guides/authentication.html
* https://docs.directus.io/guides/migrating-from-v8/
* https://github.com/directus-community/migration-tool/blob/94d66bcf595b1b6fa374669056dbe3c2791d218a/tasks/users.js#L59 

**Conclusion 3: At some point, bcrypt was supported**
I can't find where the code went, but there was a commit in summer 2020 to enable checking password hashes against bcrypt. 
* https://github.com/directus/directus/commit/d54d377a82194e52c454d8c2394181121b6a6f59

There are clearly others experiencing this issue too (#4413), and variations of the same (#4335). Although I don't think it makes sense to allow failovers for all possible hashing algorithms, bcrypt is pretty popular - and particularly I think people would expect it to work because it's what was used in v8.

I definitely like your second suggestion. Potentially this could also be optionally enabled through an environment variable (eg: BCRYPT_MIGRATE) for additionally security.

https://github.com/directus/directus/blob/43f4e63179b7c370ceee721c0a5ca0f616f30c58/api/src/services/authentication.ts#L89

**Reply by x-EricH-x** [reply]
Thanks for the very helpful info
( and impressed by how deep you dug into the hole : D )

**Reply by rijkvanzanten** [reply]
Conclusion is correct!
Conclusion 2 is also correct!
Conclusion 3 is also correct! 

The very first alpha versions of v9 still used bcrypt. It was swapped to argon2 later as it's more secure, and generally seems to be a better choice for new projects. (This SO answer has a nice detailed summary: https://security.stackexchange.com/a/197550) 

@x-EricH-x I agree that 

> alternatively, an automatic re-hash of password using Argon2 upon login if the password is hashed using bcrypt

Sounds like the best case scenario, but (un)fortunately, the passwords are saved as a _hash_, not as an encrypted string. This means that there's no way to get the plain-text password back from the bcrypt hash, and thus no way to re-hash it using argon2. 

Ultimately, the best thing we can do is to notify people of this fact in the migration guide, as per @joeinnes's recommendation: 

> they imply that you can just run the script and your Directus will be upgraded. It is true that the users are migrated, but they will not be able to log in **without resetting their password**)

_(emphasis mine)_

**Reply by joeinnes** [reply]
Although it's not possible get the plaintext back from the bcrypt, theoretically, it *would be possible* to get it from the user. We could hash the plaintext string that is checked against the bcrypt hash:

````
if ((await argon2.verify(user.password, password)) === false) { 
  if  ((await bcrypt.verify(user.password, password)) === true) {
    const newHash = argon2.hash(password);
    // Persist the new password to the database
  }
}
````

Frankly, not sure if this is a sensible thing to do, as it does seem to potentially open up a new attack vector, but throwing it out there for discussion.

**Reply by rijkvanzanten** [reply]
True, true!

Not sure if that technical overhead and potential security situation is worth the backwards compatibility though 🤔 

Asking the end user to reset their password once after a major upgrade doesn't feel like too much to ask (to me, right now)

**Reply by x-EricH-x** [reply]
Just wonder, isn't the login API already taking the plain password ([link](https://docs.directus.io/reference/api/rest/authentication/#login))?

I was under the impression that we can rehash using it (plain password input), am I missing something?

**Reply by joeinnes** [reply]
Short answer is that yes, that's possible, and if you want to implement something like that yourself, it's not too complicated (see above, although I didn't immediately find how you can persist a password hash to the database without initiating a reset email, it must be possible because it can be done from the user page on the app).

However, having this enabled by Directus as part of the core platform introduces potential security risks, and negates the benefits of choosing Argon2 over bcrypt in the first place.

There are a few approaches you could take

1. Try to upgrade the hashes in the database before migration - pro: seamless migration for end users, con: you'll end up with some users not logging in so you'll not achieve 100% coverage, likely an extended timeline, and you'll likely have to work out the Argon2 integration into the previous CMS yourself

2. As said above, you could try to silently rehash the passwords on login when users first access the new system - pro: seamless migration for end users, con: potentially opening yourself up to a security risk, implementing this yourself in Directus (or paying someone to implement this for you)

3. Force users to reset their passwords the first time they log in to the new platform - pros: you force users to revalidate their email addresses, no increase in attack surface, con: slight inconvenience to your users.

**Comment by pierre-H** [comment]
I actually did that in the V8 version in order to migrate from an old algo to bcrypt.
I was accessing the password by using the `request` service.

Now, it's not even possible to do it ourselves because there is no `request` service and the hook payload doesn't contain the password 😢  :
https://github.com/directus/directus/blob/bc30ee2bdeaa92497669c0079b6b8c9650db7ff7/api/src/services/authentication.ts#L62

For us, it's not a big deal to reset a password.
But we have a client who work with patients with rarer diseases and their parents and they are not all very comfortable with computers.
Hasking them to reset their password discourage them to continue.

**Reply by pierre-H** [reply]
The only solution I imagine is to create a custom login controller.
It will work only for the api, not for the app.

**Reply by rijkvanzanten** [reply]
You'll be able to do this very soon. @joselcvarela already gracefully provided a PR that allows you to read the full user object (including password) in the auth hooks, which will be merged asap: https://github.com/directus/directus/pull/5251

**Comment by jaads** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
